### PR TITLE
Add samplers and more evaluation modules

### PR DESCRIPTION
**Pull request by AgatheSenellart**
After this pull request it will be possible to:

- Use a different sampling strategy in the latent space. The samplers' class are stored in multivae.samplers. 

```python
from multivae.samplers.gaussian_mixture import (
    GaussianMixtureSampler,
    GaussianMixtureSamplerConfig,
)
from multivae.samplers.iaf_sampler import IAFSampler, IAFSamplerConfig

sampler_config = GaussianMixtureSamplerConfig(n_components = 10)
sampler = GaussianMixtureSampler(your_model, sampler_config)

# The sampler can be fit with:
sampler.fit(training_dataset)

# After that the sampler can be used to sample from the latent space
output_sampler = sampler.sample(100)

# The output is a ModelOutput instance with the same fields as returned by the encode function of each model 
# or the generate_from_prior function

z = output_sampler.z

# it can be fed to the decode function 
reconstruct = model.decode(output_sampler)

```

- The samplers can be used in the CoherenceEvaluator and the FIDEvaluator
```python

# Pass the sampler as input during evaluator init:
        evaluator = CoherenceEvaluator(
            model=model,
            classifiers=classifiers,
            output=path,
            test_dataset=dataset,
            eval_config=config,
            sampler=sampler,
        )
```

- Additional evaluation classes are added: a visualisation class to generate conditional and unconditional samples and log them to wandb

```python

        module = Visualization(model=model, output=path, test_dataset=dataset)
```

- A clustering class that computes k-means clustering on the training embedding and then computes an accuracy for the test_set

```python

module = Clustering(
            model=model,
            output=path,
            test_dataset=dataset,
            train_dataset=dataset,
        )
```

- This PR also fixes a few bugs.

### CODE REVIEW

File: src/multivae/metrics/fids/fids.py [hunk 1]
```diff
@@ -224,35 +242,6 @@ def compute_fid_from_conditional_generation(self, subset, gen_mod):
...
-
-        self.logger.info(f"Mean subsets FDs for mod {gen_mod} :")
-        for i in range(len(mean_accs)):
-            self.logger.info(
-                f"Given {i+1} modalities : {mean_accs[i]} +- {std_accs[i]}"
-            )
-
-        return ModelOutput(mean_accs=mean_accs, std_accs=std_accs)
-
     def mvtcae_reproduce_fids(self, gen_mod):
```
Comment (clementchadebec) [inline src/multivae/metrics/fids/fids.py:250]: Should we rename this function?

File: src/multivae/models/jnf/jnf_model.py [hunk 2]
```diff
@@ -419,8 +421,6 @@ def sample_from_poe_subset(
             ax.plot(pos[:, 0], pos[:, 1])
             ax.quiver(pos[:, 0], pos[:, 1], grad[:, 0], grad[:, 1])
 
-            # plt.savefig('monitor_hmc.png')
-        # 1/0
         # print(acc_nbr[:10] / mcmc_steps)
```
Comment (clementchadebec) [inline src/multivae/models/jnf/jnf_model.py:424]: To be removed

**Review by clementchadebec — COMMENTED** [review]
Thank you for this huge PR @AgatheSenellart !

Important feedback:
- I get an error when trying to download the models from HF hub. See the comments in the PR.
- Maybe we also need to include the samplers in the documentation

File: examples/tutorial_notebooks/computing_visualization_and_metrics.ipynb
[the remark this answers is not in the document]
Comment (AgatheSenellart) [reply examples/tutorial_notebooks/computing_visualization_and_metrics.ipynb:1]: Hi @clementchadebec! 
No I've never had this error. I will the notebooks in a other setting than the one I am using, to see if it works as well

[the remark this answers is not in the document]
Comment (AgatheSenellart) [reply examples/tutorial_notebooks/computing_visualization_and_metrics.ipynb:1]: I wonder if you don't need to have the class we are trying to unpickle defined somewhere in the repository and if this is  what may cause this error.

[the remark this answers is not in the document]
Comment (AgatheSenellart) [reply examples/tutorial_notebooks/computing_visualization_and_metrics.ipynb:1]: I have ran the notebook on another computer and it seems fine. What python version do you have?

[the remark this answers is not in the document]
Comment (clementchadebec) [reply examples/tutorial_notebooks/computing_visualization_and_metrics.ipynb:1]: I am using python Python 3.8.16

[the remark this answers is not in the document]
Comment (clementchadebec) [reply examples/tutorial_notebooks/computing_visualization_and_metrics.ipynb:1]: It was indeed an issue linked to the Python version.

I have added an `environment.json` file that is saved when calling `save` for the model. If the model is saved using `3.9+` and download in python `3.8` env . The following is raised (and conversely)
```
http.cookiejar.LoadError: Trying to reload a model saved with python3.9+ with python3.8. Please create a virtual env with python 3.9+ to reload this model.
```

**Review by clementchadebec — APPROVED** [review]

**Pull request merged**